**About anomaly dection exercises** 
Exercise

1. Take a few minutes and jot down some example applications, specific problems, or domains where detecting anomalies is super valuable and impactful.
2. Identify a few example applications, specific problems, or domains where the outliers can skew the data in a negative way and those outliers may be safely avoided for producing a model.

1. reading MRI, fraud alert, cybersecurity

2. temperatures measuring - Kelvin degrees, zip codes with more than 5 digits (for US addresses), luxury properties in S Cal Zillow dataset

# continuous_probabilistic_methods EXERCISES

1. Define a function named get_lower_and_upper_bounds that has two arguments. The first argument is a pandas Series. The second argument is the multiplier, which should have a default argument of 1.5.


1. Using lemonade.csv dataset and focusing on continuous variables:

- Use the IQR Range Rule and the upper and lower bounds to identify the lower outliers of each column of lemonade.csv, using the multiplier of 1.5. Do these lower outliers make sense?Which outliers should be kept?
- Use the IQR Range Rule and the upper and lower bounds to identify the upper outliers of each column of lemonade.csv, using the multiplier of 1.5. Do these lower outliers make sense?Which outliers should be kept?
- Using the multiplier of 3, IQR Range Rule, and the lower and upper bounds, identify the outliers below the lower bound in each colum of lemonade.csv. Do these lower outliers make sense?Which outliers should be kept?
- Using the multiplier of 3, IQR Range Rule, and the lower and upper bounds, identify the outliers above the upper_bound in each colum of lemonade.csv. Do these upper outliers make sense? Which outliers should be kept?

2. Identify if any columns in lemonade.csv are normally distributed. For normally distributed columns:

- Use a 2 sigma decision rule to isolate the outliers.
    - Do these make sense?
    - Should certain outliers be kept or removed?
- Now use a 3 sigma decision rule to isolate the outliers in the normally distributed columns from lemonade.csv


In [1]:
import prepare
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


In [2]:
# 1. upper and lower bound functions 
#mild_upper = prepare.get_upper_outliers(s, k)
#mild_lower = prepare.get_lower_outliers(s, k)

In [3]:
# df = pd.read_clipboard(sep=',')
# df.head()

In [4]:
#df.to_csv('lemonade.csv')

In [5]:
df = pd.read_csv('lemonade.csv', index_col=0)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 365 entries, 0 to 364
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Date         365 non-null    object 
 1   Day          365 non-null    object 
 2   Temperature  365 non-null    float64
 3   Rainfall     365 non-null    float64
 4   Flyers       365 non-null    int64  
 5   Price        365 non-null    float64
 6   Sales        365 non-null    int64  
dtypes: float64(3), int64(2), object(2)
memory usage: 22.8+ KB


**1. get upper and lower outliers for each column** set multiplier to 1.5

In [7]:
upperdf = prepare.add_upper_outlier_columns(df, k=1.5)
upperdf.head()

Date        Day  Temperature  Rainfall  Flyers  Price  Sales  \
0  1/1/17     Sunday         27.0      2.00      15    0.5     10   
1  1/2/17     Monday         28.9      1.33      15    0.5     13   
2  1/3/17    Tuesday         34.5      1.33      27    0.5     15   
3  1/4/17  Wednesday         44.1      1.05      28    0.5     17   
4  1/5/17   Thursday         42.4      1.00      33    0.5     18   

   Temperature_up_outliers  Rainfall_up_outliers  Flyers_up_outliers  \
0                      0.0                  0.70                 0.0   
1                      0.0                  0.03                 0.0   
2                      0.0                  0.03                 0.0   
3                      0.0                  0.00                 0.0   
4                      0.0                  0.00                 0.0   

   Price_up_outliers  Sales_up_outliers  
0                0.0                0.0  
1                0.0                0.0  
2                0.0                0.0  
3                0.0                0.0  
4                0.0                0.0

In [8]:
lowerdf = prepare.add_lower_outlier_columns(df, k=1.5)
lowerdf.head()

Date        Day  Temperature  Rainfall  Flyers  Price  Sales  \
0  1/1/17     Sunday         27.0      2.00      15    0.5     10   
1  1/2/17     Monday         28.9      1.33      15    0.5     13   
2  1/3/17    Tuesday         34.5      1.33      27    0.5     15   
3  1/4/17  Wednesday         44.1      1.05      28    0.5     17   
4  1/5/17   Thursday         42.4      1.00      33    0.5     18   

   Temperature_low_outliers  Rainfall_low_outliers  Flyers_low_outliers  \
0                       0.0                      0                  0.0   
1                       0.0                      0                  0.0   
2                       0.0                      0                  0.0   
3                       0.0                      0                  0.0   
4                       0.0                      0                  0.0   

   Price_low_outliers  Sales_low_outliers  
0                 0.0                   0  
1                 0.0                   0  
2                 0.0                   0  
3                 0.0                   0  
4                 0.0                   0

In [9]:
df.describe()

Temperature    Rainfall      Flyers  Price       Sales
count   365.000000  365.000000  365.000000  365.0  365.000000
mean     61.224658    0.825973   40.104110    0.5   27.865753
std      18.085892    0.273890   13.786445    0.0   30.948132
min      15.100000    0.400000  -38.000000    0.5    7.000000
25%      49.700000    0.650000   31.000000    0.5   20.000000
50%      61.100000    0.740000   39.000000    0.5   25.000000
75%      71.700000    0.910000   49.000000    0.5   30.000000
max     212.000000    2.500000   80.000000    0.5  534.000000

**1. do the outliers make sense?** at 1.5 multiplier

- UPPER OUTLIERS
    - Temperature: 1 observation = 212 degrees, YES this is outlier/typo
    - Rainfall: 28 observations, NO - high rain in Dec and Jan makes sense
    - Flyers: 2 observations, MAYBE, could just be that these 2 Fridays decision was made to focus on flyers
    - Price: none
    - Sales: 4 observations, NO - these are all leading up to July 4th so make sense related to high sales

- LOWER OUTLIERS
    - Temperature: 1 observation = 15.4 degrees, NO this is only slightly colder than lower bound of 16.7
    - Rainfall: none
    - Flyers: 1 observation = -38, YES, typo, can not have negative flyers handed out
    - Price: none
    - Sales: none

In [10]:
upperdf[upperdf.Temperature_up_outliers != 0]

Date       Day  Temperature  Rainfall  Flyers  Price  Sales  \
41  2/11/17  Saturday        212.0      0.91      35    0.5     21   

    Temperature_up_outliers  Rainfall_up_outliers  Flyers_up_outliers  \
41                    107.3                   0.0                 0.0   

    Price_up_outliers  Sales_up_outliers  
41                0.0                0.0

In [11]:
upperdf[upperdf.Rainfall_up_outliers != 0]

Date        Day  Temperature  Rainfall  Flyers  Price  Sales  \
0      1/1/17     Sunday         27.0      2.00      15    0.5     10   
1      1/2/17     Monday         28.9      1.33      15    0.5     13   
2      1/3/17    Tuesday         34.5      1.33      27    0.5     15   
5      1/6/17     Friday         25.3      1.54      23    0.5     11   
6      1/7/17   Saturday         32.9      1.54      19    0.5     13   
10    1/11/17  Wednesday         32.6      1.54      23    0.5     12   
11    1/12/17   Thursday         38.2      1.33      16    0.5     14   
12    1/13/17     Friday         37.5      1.33      19    0.5     15   
15    1/16/17     Monday         30.6      1.67      24    0.5     12   
16    1/17/17    Tuesday         32.2      1.43      26    0.5     14   
19    1/20/17     Friday         31.6      1.43      20    0.5     12   
23    1/24/17    Tuesday         28.6      1.54      20    0.5     12   
27    1/28/17   Saturday         34.9      1.33      15    0.5     13   
28    1/29/17     Sunday         35.2      1.33      27    0.5     14   
337   12/4/17     Monday         34.9      1.54      16    0.5     13   
338   12/5/17    Tuesday         22.0      1.82      11    0.5     10   
342   12/9/17   Saturday         31.2      1.43      19    0.5     14   
343  12/10/17     Sunday         31.3      1.82      15    0.5     11   
345  12/12/17    Tuesday         33.5      1.33      22    0.5     15   
346  12/13/17  Wednesday         32.2      1.43      26    0.5     14   
347  12/14/17   Thursday         31.9      1.54      24    0.5     13   
350  12/17/17     Sunday         32.2      1.33      16    0.5     14   
351  12/18/17     Monday         30.9      1.43      27    0.5     13   
354  12/21/17   Thursday         40.5      1.33      23    0.5     15   
355  12/22/17     Friday         30.9      1.54      17    0.5     13   
359  12/26/17    Tuesday         28.9      1.43      23    0.5     13   
363  12/30/17   Saturday         30.9      1.43      22    0.5     13   
364  12/31/17     Sunday         15.1      2.50       9    0.5      7   

     Temperature_up_outliers  Rainfall_up_outliers  Flyers_up_outliers  \
0                        0.0                  0.70                 0.0   
1                        0.0                  0.03                 0.0   
2                        0.0                  0.03                 0.0   
5                        0.0                  0.24                 0.0   
6                        0.0                  0.24                 0.0   
10                       0.0                  0.24                 0.0   
11                       0.0                  0.03                 0.0   
12                       0.0                  0.03                 0.0   
15                       0.0                  0.37                 0.0   
16                       0.0                  0.13                 0.0   
19                       0.0                  0.13                 0.0   
23                       0.0                  0.24                 0.0   
27                       0.0                  0.03                 0.0   
28                       0.0                  0.03                 0.0   
337                      0.0                  0.24                 0.0   
338                      0.0                  0.52                 0.0   
342                      0.0                  0.13                 0.0   
343                      0.0                  0.52                 0.0   
345                      0.0                  0.03                 0.0   
346                      0.0                  0.13                 0.0   
347                      0.0                  0.24                 0.0   
350                      0.0                  0.03                 0.0   
351                      0.0                  0.13                 0.0   
354                      0.0                  0.03                 0.0   
355                      0.0             

In [12]:
upperdf[upperdf.Flyers_up_outliers != 0]

Date     Day  Temperature  Rainfall  Flyers  Price  Sales  \
166  6/16/17  Friday         99.3      0.47      77    0.5     41   
194  7/14/17  Friday         92.0      0.50      80    0.5     40   

     Temperature_up_outliers  Rainfall_up_outliers  Flyers_up_outliers  \
166                      0.0                   0.0                 1.0   
194                      0.0                   0.0                 4.0   

     Price_up_outliers  Sales_up_outliers  
166                0.0                0.0  
194                0.0                0.0

In [13]:
upperdf[upperdf.Price_up_outliers != 0]

Empty DataFrame
Columns: [Date, Day, Temperature, Rainfall, Flyers, Price, Sales, Temperature_up_outliers, Rainfall_up_outliers, Flyers_up_outliers, Price_up_outliers, Sales_up_outliers]
Index: []

In [14]:
upperdf[upperdf.Sales_up_outliers != 0]


Date       Day  Temperature  Rainfall  Flyers  Price  Sales  \
181  7/1/17  Saturday        102.9      0.47      59    0.5    143   
182  7/2/17    Sunday         93.4      0.51      68    0.5    158   
183  7/3/17    Monday         81.5      0.54      68    0.5    235   
184  7/4/17   Tuesday         84.2      0.59      49    0.5    534   

     Temperature_up_outliers  Rainfall_up_outliers  Flyers_up_outliers  \
181                      0.0                   0.0                 0.0   
182                      0.0                   0.0                 0.0   
183                      0.0                   0.0                 0.0   
184                      0.0                   0.0                 0.0   

     Price_up_outliers  Sales_up_outliers  
181                0.0               98.0  
182                0.0              113.0  
183                0.0              190.0  
184                0.0              489.0

In [15]:
lowerdf[lowerdf.Temperature_low_outliers != 0]

Date     Day  Temperature  Rainfall  Flyers  Price  Sales  \
364  12/31/17  Sunday         15.1       2.5       9    0.5      7   

     Temperature_low_outliers  Rainfall_low_outliers  Flyers_low_outliers  \
364                       1.6                      0                  0.0   

     Price_low_outliers  Sales_low_outliers  
364                 0.0                   0

In [19]:
lowerdf[lowerdf.Rainfall_low_outliers != 0]

Empty DataFrame
Columns: [Date, Day, Temperature, Rainfall, Flyers, Price, Sales, Temperature_low_outliers, Rainfall_low_outliers, Flyers_low_outliers, Price_low_outliers, Sales_low_outliers]
Index: []

In [20]:
lowerdf[lowerdf.Flyers_low_outliers != 0]

Date      Day  Temperature  Rainfall  Flyers  Price  Sales  \
324  11/21/17  Tuesday         47.0      0.95     -38    0.5     20   

     Temperature_low_outliers  Rainfall_low_outliers  Flyers_low_outliers  \
324                       0.0                      0                 42.0   

     Price_low_outliers  Sales_low_outliers  
324                 0.0                   0

In [21]:
lowerdf[lowerdf.Price_low_outliers != 0]

Empty DataFrame
Columns: [Date, Day, Temperature, Rainfall, Flyers, Price, Sales, Temperature_low_outliers, Rainfall_low_outliers, Flyers_low_outliers, Price_low_outliers, Sales_low_outliers]
Index: []

In [22]:
lowerdf[lowerdf.Sales_low_outliers != 0]

Empty DataFrame
Columns: [Date, Day, Temperature, Rainfall, Flyers, Price, Sales, Temperature_low_outliers, Rainfall_low_outliers, Flyers_low_outliers, Price_low_outliers, Sales_low_outliers]
Index: []

1. get upper and lower outliers for each column set multiplier to 3

In [23]:
upperdf3 = prepare.add_upper_outlier_columns(df, k=3)
upperdf3.head()

Date        Day  Temperature  Rainfall  Flyers  Price  Sales  \
0  1/1/17     Sunday         27.0      2.00      15    0.5     10   
1  1/2/17     Monday         28.9      1.33      15    0.5     13   
2  1/3/17    Tuesday         34.5      1.33      27    0.5     15   
3  1/4/17  Wednesday         44.1      1.05      28    0.5     17   
4  1/5/17   Thursday         42.4      1.00      33    0.5     18   

   Temperature_up_outliers  Rainfall_up_outliers  Flyers_up_outliers  \
0                      0.0                  0.31                   0   
1                      0.0                  0.00                   0   
2                      0.0                  0.00                   0   
3                      0.0                  0.00                   0   
4                      0.0                  0.00                   0   

   Price_up_outliers  Sales_up_outliers  
0                0.0                0.0  
1                0.0                0.0  
2                0.0                0.0  
3                0.0                0.0  
4                0.0                0.0

In [24]:
lowerdf3 = prepare.add_lower_outlier_columns(df, k=3)
lowerdf3.head()

Date        Day  Temperature  Rainfall  Flyers  Price  Sales  \
0  1/1/17     Sunday         27.0      2.00      15    0.5     10   
1  1/2/17     Monday         28.9      1.33      15    0.5     13   
2  1/3/17    Tuesday         34.5      1.33      27    0.5     15   
3  1/4/17  Wednesday         44.1      1.05      28    0.5     17   
4  1/5/17   Thursday         42.4      1.00      33    0.5     18   

   Temperature_low_outliers  Rainfall_low_outliers  Flyers_low_outliers  \
0                         0                      0                  0.0   
1                         0                      0                  0.0   
2                         0                      0                  0.0   
3                         0                      0                  0.0   
4                         0                      0                  0.0   

   Price_low_outliers  Sales_low_outliers  
0                 0.0                   0  
1                 0.0                   0  
2                 0.0                   0  
3                 0.0                   0  
4                 0.0                   0

**1. do the outliers make sense?** at 3 multiplier

- UPPER OUTLIERS
    - Temperature: 1 observation, YES, 212 degrees must be a typo
    - Rainfall: 4 observations, 1 in Jan, 3 in Dec, higher than average for the dataset, but not unreasonable, would keep these
    - Flyers: none
    - Price: none
    - Sales: 4 observations leading up and on July 4th, depending on location this is reasonable, would keep these

- LOWER OUTLIERS
    - Temperature: none
    - Rainfall: none
    - Flyers: 1 observation, YES, can not have negative flyers handed out
    - Price: none
    - Sales: none

In [25]:
upperdf3[upperdf3.Temperature_up_outliers != 0]

Date       Day  Temperature  Rainfall  Flyers  Price  Sales  \
41  2/11/17  Saturday        212.0      0.91      35    0.5     21   

    Temperature_up_outliers  Rainfall_up_outliers  Flyers_up_outliers  \
41                     74.3                   0.0                   0   

    Price_up_outliers  Sales_up_outliers  
41                0.0                0.0

In [26]:
upperdf3[upperdf3.Rainfall_up_outliers != 0]

Date      Day  Temperature  Rainfall  Flyers  Price  Sales  \
0      1/1/17   Sunday         27.0      2.00      15    0.5     10   
338   12/5/17  Tuesday         22.0      1.82      11    0.5     10   
343  12/10/17   Sunday         31.3      1.82      15    0.5     11   
364  12/31/17   Sunday         15.1      2.50       9    0.5      7   

     Temperature_up_outliers  Rainfall_up_outliers  Flyers_up_outliers  \
0                        0.0                  0.31                   0   
338                      0.0                  0.13                   0   
343                      0.0                  0.13                   0   
364                      0.0                  0.81                   0   

     Price_up_outliers  Sales_up_outliers  
0                  0.0                0.0  
338                0.0                0.0  
343                0.0                0.0  
364                0.0                0.0

In [27]:
upperdf3[upperdf3.Flyers_up_outliers != 0]

Empty DataFrame
Columns: [Date, Day, Temperature, Rainfall, Flyers, Price, Sales, Temperature_up_outliers, Rainfall_up_outliers, Flyers_up_outliers, Price_up_outliers, Sales_up_outliers]
Index: []

In [28]:
upperdf3[upperdf3.Price_up_outliers != 0]

Empty DataFrame
Columns: [Date, Day, Temperature, Rainfall, Flyers, Price, Sales, Temperature_up_outliers, Rainfall_up_outliers, Flyers_up_outliers, Price_up_outliers, Sales_up_outliers]
Index: []

In [29]:
upperdf3[upperdf3.Sales_up_outliers != 0]

Date       Day  Temperature  Rainfall  Flyers  Price  Sales  \
181  7/1/17  Saturday        102.9      0.47      59    0.5    143   
182  7/2/17    Sunday         93.4      0.51      68    0.5    158   
183  7/3/17    Monday         81.5      0.54      68    0.5    235   
184  7/4/17   Tuesday         84.2      0.59      49    0.5    534   

     Temperature_up_outliers  Rainfall_up_outliers  Flyers_up_outliers  \
181                      0.0                   0.0                   0   
182                      0.0                   0.0                   0   
183                      0.0                   0.0                   0   
184                      0.0                   0.0                   0   

     Price_up_outliers  Sales_up_outliers  
181                0.0               83.0  
182                0.0               98.0  
183                0.0              175.0  
184                0.0              474.0

In [30]:
lowerdf3[lowerdf3.Temperature_low_outliers != 0]

Empty DataFrame
Columns: [Date, Day, Temperature, Rainfall, Flyers, Price, Sales, Temperature_low_outliers, Rainfall_low_outliers, Flyers_low_outliers, Price_low_outliers, Sales_low_outliers]
Index: []

In [31]:
lowerdf3[lowerdf3.Rainfall_low_outliers != 0]

Empty DataFrame
Columns: [Date, Day, Temperature, Rainfall, Flyers, Price, Sales, Temperature_low_outliers, Rainfall_low_outliers, Flyers_low_outliers, Price_low_outliers, Sales_low_outliers]
Index: []

In [32]:
lowerdf3[lowerdf3.Flyers_low_outliers != 0]

Date      Day  Temperature  Rainfall  Flyers  Price  Sales  \
324  11/21/17  Tuesday         47.0      0.95     -38    0.5     20   

     Temperature_low_outliers  Rainfall_low_outliers  Flyers_low_outliers  \
324                         0                      0                 15.0   

     Price_low_outliers  Sales_low_outliers  
324                 0.0                   0

In [33]:
lowerdf3[lowerdf3.Price_low_outliers != 0]

Empty DataFrame
Columns: [Date, Day, Temperature, Rainfall, Flyers, Price, Sales, Temperature_low_outliers, Rainfall_low_outliers, Flyers_low_outliers, Price_low_outliers, Sales_low_outliers]
Index: []

In [34]:
lowerdf3[lowerdf3.Sales_low_outliers != 0]

Empty DataFrame
Columns: [Date, Day, Temperature, Rainfall, Flyers, Price, Sales, Temperature_low_outliers, Rainfall_low_outliers, Flyers_low_outliers, Price_low_outliers, Sales_low_outliers]
Index: []

In [ ]:
Identify if any columns in lemonade.csv are normally distributed.